src/runtime_src/tools/xclbinutil/KernelUtilities.cxx

```diff
@@ -306,26 +322,67 @@ void addArgsToMemoryConnections(const unsigned int ipLayoutIndexID,
         ptMemData.put("m_base_address", "0x0");
         memTopology.push_back(ptMemData);
       }
-
-      memoryConnection = "MEM_PS_KERNEL";
-    }
-
-    // Do we have a connection to perform?
-    if (memoryConnection != NOT_DEFINED) {
-      auto iter = std::find_if(memTopology.begin(), memTopology.end(),
+      // memoryConnection = "MEM_PS_KERNEL";
+
+      // the newly added mem bank is always at the last of memTopology
+      const unsigned int memIndex = static_cast<unsigned int>(memTopology.size()) - 1;
+      vMemBanks.push_back(std::to_string(memIndex));
+    } else {
+      // memory-connection could be
+      //   a comma separated memory bank indices, for --add-pskernel
+      //   a string which matches to m_tag, for --add-kernel
+      std::vector<int> viMemBanks;
+      auto isNumeric = [&viMemBanks](const std::string& str) {
+        std::istringstream ss(str);
+        std::string token;
+        while(std::getline(ss, token, ',')) {
+          int index;
+          try {
+            index = std::stoi(token);
+          } catch (const std::exception&) {
+            return false;
+          }
+          viMemBanks.push_back(index);
+        }
+        return true;
+      };
+
+      if (isNumeric(memoryConnection)) {
+        // make sure the specified index exists
+        for (const auto& memBankIndex : viMemBanks) {
+          if (memBankIndex >= 0 && memBankIndex < static_cast<int>(memTopology.size())) {
+            vMemBanks.push_back(std::to_string(memBankIndex));
+            std::cout << "\tconnecting arg " << std::to_string(argIndexID) << " to mem bank " << std::to_string(memBankIndex) << std::endl;
+          }
+          else
+            throw std::runtime_error("Specifїed mem_banks is not valid");
```

Comment:
This error message should be a bit more descriptive.  For example:
```suggestion
         {
            const auto errMsg = boost::format("Specified memory bank (%d) is invalid.  Valid ranges are from 0 to %d.") % memBankIndex % memTopology.size(); 
            throw std::runtime_error(errMsg.str());
         }
```

Reply:
message udpated